qiskit_experiments/framework/backend_timing_mixin.py

```diff
@@ -0,0 +1,172 @@
+
+        Raises:
+            QiskitError: The backend is not set or does not include a dt value.
+        """
+        if self.backend is not None and hasattr(self.backend.configuration(), "dt"):
+            return self.backend.configuration().dt
+
+        raise QiskitError("Backend must be set to consider sample timing.")
+
+    def _delay_duration(self, time: float, next_instruction: str = "pulse") -> Union[int, float]:
+        """The delay duration closest to ``time`` in delay units
+
+        This method produces the value to pass for the ``duration`` of a
+        ``Delay`` instruction of a ``QuantumCircuit`` so that the delay fills
+        the time until the next pulse.  This method does a little more than
+        just convert ``time`` to the closest number of samples. The pulse
+        timing constraints of the backend are considered in order to give a
+        number of samples closest to ``time`` plus however many more samples
+        are needed to get to the next valid sample for the start of a pulse in
+        a subsequent instruction.
+
+        This method is useful when it is desired that a delay instruction
+        accurately reflect how long the hardware will pause between two other
+        instructions.
+
+        Args:
+            time: The nominal delay time to convert
+            next_instruction: Either "pulse", "acquire", or "both" to indicate
+                whether the next instruction will be a pulse, an acquire
+                instruction, or both.
+
+        Returns:
+            The delay duration in samples or seconds, depending on the value of
+            :meth:`.BackendTimingMixin._delay_unit`.
+
+        Raises:
+            QiskitError: Bad value for ``next_instruction``.
+        """
+        if self._delay_unit == "s":
+            return time
+
+        timing_constraints = getattr(self.backend.configuration(), "timing_constraints", {})
+        pulse_alignment = timing_constraints.get("pulse_alignment", 1)
+        acquire_alignment = timing_constraints.get("acquire_alignment", 1)
+
+        if next_instruction == "both":
+            # We round the delay values to the least common multiple of the
+            # alignment constraints so that the delay can fill all the time
+            # until a subsequent pulse satisfying the alignment constraints.
+            #
+            # Replace with math.lcm(pulse_alignment, acquire_alignment) when
+            # dropping support for Python 3.8
+            granularity = (
+                pulse_alignment * acquire_alignment // math.gcd(pulse_alignment, acquire_alignment)
+            )
+        elif next_instruction == "pulse":
+            granularity = pulse_alignment
+        elif next_instruction == "acquire":
+            granularity = acquire_alignment
+        else:
+            raise QiskitError(f"Bad value for next_instruction: {next_instruction}")
+
+        samples = int(round(time / self._dt / granularity) * granularity)
```

Comment:
I just notice this is not always correct. For example, when `pulse_alignment=1` and `acquire_alignment=16`,

```python
qc = QuantumCircuit(1, 1)
qc.sx(0)
qc.delay(self._delay_duration(5), 0)
qc.x(0)
qc.delay(self._delay_duration(16, "acquire"), 0)
qc.measure(0, 0)
```
This creates a circuit with two delays, i.e. 5dt-x-16-dt. However, the actual t0 of the acquire instruction will be not a multiple of 16 which will give you unexpected outcome (frame might be off).

Reply:
To be on the safe side we can always use `both` option.

Reply:
Hmm, this is very subtle. I thought the backend used as late as possible scheduling, so if a measurement was done last it would always be lined up with the acquisition (if the measurement instruction duration was a multiple of the alignment value), but it seems like that is not the case for a test I tried. On a backend with the alignment values you gave (ibm_cairo). I ran these two similar sets of circuits:

```
qcs = []
for delay in range(16):
    qc = QuantumCircuit(1, 1)
    qc.x(0)
    qc.delay(delay, 0)
    qc.measure(0, 0)
    qcs.append(qc)
    
job = backend.run(qcs, shots=250)
```

and

```
qcs = []
for delay in range(16):
    qc = QuantumCircuit(1, 1)
    qc.delay(16 - delay, 0)
    qc.x(0)
    qc.delay(delay, 0)
    qc.measure(0, 0)
    qcs.append(qc)
    
job = backend.run(qcs, shots=250)
```

The second version prepads a delay with the balance of the acquire_alignment so that the measure always occurs at the same time when the circuit is scheduled as a `Schedule` (though the backend does not actually convert to a `Schedule` when running on hardware).

The results were:

```
{'0x0': 6, '0x1': 244}
{'0x0': 2, '0x1': 248}
{'0x0': 1, '0x1': 249}
{'0x0': 1, '0x1': 249}
{'0x0': 1, '0x1': 249}
{'0x0': 2, '0x1': 248}
{'0x0': 3, '0x1': 247}
{'0x0': 14, '0x1': 236}
{'0x0': 41, '0x1': 209}
{'0x0': 89, '0x1': 161}
{'0x0': 162, '0x1': 88}
{'0x0': 206, '0x1': 44}
{'0x0': 229, '0x1': 21}
{'0x0': 239, '0x1': 11}
{'0x0': 247, '0x1': 3}
{'0x0': 249, '0x1': 1}
```

and

```
{'0x0': 4, '0x1': 246}
{'0x0': 4, '0x1': 246}
{'0x0': 3, '0x1': 247}
{'0x0': 4, '0x1': 246}
{'0x0': 4, '0x1': 246}
{'0x0': 3, '0x1': 247}
{'0x0': 1, '0x1': 249}
{'0x0': 2, '0x1': 248}
{'0x0': 3, '0x1': 247}
{'0x0': 6, '0x1': 244}
{'0x0': 4, '0x1': 246}
{'0x0': 4, '0x1': 246}
{'0x0': 2, '0x1': 248}
{'0x0': 4, '0x1': 246}
{'0x0': 5, '0x1': 245}
{'0x0': 4, '0x1': 246}
```

So without pre-padding, the measured phase is getting rotated.

So for now should we just drop `next_instruction` and always use the lcm? Since pretty much all circuits want to measure at the end and instructions are getting pushed sooner, every drive pulse impacts the timing of the measurement of its qubit. As long as the numbers are round, it is easy to use the lcm of the alignments. I could see a future where it would be useful to have the measurement get pre-padded with whatever delay was needed to line up with the acquisition while still having freedom to vary delays on drive channels -- like a Ramsey experiment with very different drive and acquisition sampling rates where you wouldn't care if there were a slight delay between the second qubit pulse and the measurement as long as the delay between the two qubit pulses was precise.

Reply:
Yes, I think using lcm is reasonable. Otherwise, we need to implement the logic of scheduling pass here, which is in principle the same as calling transpiler. I think the effect of rounding is quite limited, and indeed it doesn't matter as long as we use proper xval for analysis. So rounding should be better than calling transpile in terms of performance.

I think the better future is applying a counter rotation on the IQ plane in the post processing so that we can start measure instruction at arbitrary time. But this cannot be done in the circuit domain, and we always need to attach custom measure instruction as a pulse gates. It would be great backend can automatically do this.

Reply:
Yes, having the phase handled on the backend would be best. We could look into support for this for IBM backends. For varying the measurement phase to be fully supported, it would need to work for hardware state discrimination, so that would require hardware enablement, not just rotating the phase of level 1 data returned by the backend. Newer IBM hardware has `pulse_alignment == acquire_alignment` where this phase doesn't even come up so it might not be a direction that IBM wants to prioritize.

When the pulse shifts relative to the acquisition, the kernel is also technically not quite accurate but it is only off a few ns which probably doesn't matter much.

For the context of this PR, I will make lcm the only method and add a comment explaining why.

Reply:
Ah, you're right. I didn't think about kernel. That makes sense.

Reply:
I spent a long time thinking about this one and put a long discussion in the doc string.